compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -380,6 +386,191 @@ void transform_dialect::ForallToWorkgroupOp::getEffects(
+
+FailureOr<int64_t> getTripCount(scf::ForallOp loop) {
+  ArrayRef<int64_t> lbs = loop.getStaticLowerBound();
+  ArrayRef<int64_t> ubs = loop.getStaticUpperBound();
+  ArrayRef<int64_t> steps = loop.getStaticStep();
+
+  if (ShapedType::isDynamicShape(lbs) || ShapedType::isDynamicShape(ubs) ||
+      ShapedType::isDynamicShape(steps)) {
+    return failure();
+  }
+
+  int64_t tripCount = 1;
+  for (auto [lb, ub, step] : llvm::zip_equal(lbs, ubs, steps)) {
+    tripCount *= (ub - lb) / step;
+  }
+  return tripCount;
+}
+
+LogicalResult compareWorkerCountsAndTypes(scf::ForallOp producer,
+                                          scf::ForallOp consumer) {
+  FailureOr<int64_t> producerTripCount = getTripCount(producer);
+  FailureOr<int64_t> consumerTripCount = getTripCount(consumer);
+  if (failed(producerTripCount) || failed(consumerTripCount) ||
+      *producerTripCount != *consumerTripCount) {
+    return failure();
+  }
+
+  mlir::TypeID mappingId = producer.getMapping()
+                               ->getValue()
+                               .front()
+                               .getAbstractAttribute()
+                               .getTypeID();
+  auto compareMappingTypes = [&](ArrayAttr array) {
+    return llvm::all_of(array.getValue(), [&](const Attribute &attr) {
+      return attr.getAbstractAttribute().getTypeID() == mappingId;
+    });
+  };
+
+  if (!compareMappingTypes(producer.getMappingAttr()) ||
+      !compareMappingTypes(consumer.getMappingAttr())) {
+    return failure();
+  }
+  return success();
+}
+
+Value getReplacementSlice(RewriterBase &rewriter, Location loc,
+                          tensor::ParallelInsertSliceOp parallelInsert,
+                          tensor::ExtractSliceOp extractSlice) {
+  return rewriter.create<IREE::GPU::ShuffleTensorOp>(
+      loc, extractSlice.getType(), parallelInsert.getSource(),
+      parallelInsert.getOffsets(), parallelInsert.getSizes(),
+      parallelInsert.getStrides(), parallelInsert.getStaticOffsets(),
+      parallelInsert.getStaticSizes(), parallelInsert.getStaticStrides(),
+      parallelInsert.getDest(), extractSlice.getOffsets(),
+      extractSlice.getSizes(), extractSlice.getStrides(),
+      extractSlice.getStaticOffsets(), extractSlice.getStaticSizes(),
+      extractSlice.getStaticStrides());
+}
+
+LogicalResult fuseForallIntoSlice(RewriterBase &rewriter,
+                                  scf::ForallOp producer,
+                                  scf::ForallOp consumer,
+                                  tensor::ExtractSliceOp slice) {
+  if (producer->getNumResults() != 1) {
+    return failure();
+  }
+
+  if (failed(compareWorkerCountsAndTypes(producer, consumer))) {
+    return failure();
+  }
+
+  auto isAll = [](ArrayRef<int64_t> array, int64_t cmp) {
+    return llvm::all_of(array, [cmp](int64_t x) { return x == cmp; });
+  };
+
+  if (!isAll(producer.getStaticStep(), 1) ||
```

Comment:
Maybe just use the `getMixedStep` and `isConstantInt` methods to check for this? I know you have checked for static shapes earlier, but its a bit of an "action at a distance thing"